File: src/Chapter4/SystemElement.java
package Chapter4;

/*                                                                            *
 * @author Gianluca Capone & Davide Sgobba                                    *
 *                                                                            */

/* This class contains all variables that define the system element produced  *
 * by computer firms and the methods that operate over these variables        */
class SystemElement {
    
    // VARIABLES
    double mod;
    /* Merit of design of the system element (M-SY_f,t)                       */
    double muProg;
    /* Mean of the distribution of system technical progress (mu-SY_f,t)      */
    
    //TECHNICAL VARIABLES & OBJECTS
    ComputerFirm Firm;
    /* Access to the firm                                                     */
    
    // CONSTRUCTOR
    SystemElement(double MOD, ComputerFirm FIRM) {
        mod  = MOD;
        Firm = FIRM;
    }

    /* This method computes the mod of the system based on the level of past  *
     * mod and the level of public knowledge                                  */
    void calcMod() {
        
        double zmax = 0;
        /* Equation 14.a                                                      */
        muProg = Math.log(Firm.ComputerMarket.pkSys) * (1 - Firm.ComputerMarket.internalCum)
                 + Math.log(mod) * Firm.ComputerMarket.internalCum;
        for (int i = 1; i <= Firm.numOfDrawsSys; i++) {
            double z = Math.exp(muProg + Math.sqrt(Firm.ComputerMarket.sdSys)
                         * Firm.ComputerMarket.rng.nextGaussian());
            if (z > zmax) {
                zmax = z;
            }
        }
        if (zmax > mod) {
            mod = zmax;
        }
    }
    
    /* This method is activated when exit occurs: it resets the most relevant *
     * variables at the product level                                         */
    void exitSystem() {
        muProg  = 0;
        mod     = 0;
    }
}

File: src_py/Chapter4/system_element.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
SystemElement模块 - SystemElement类的Python实现
转换自Java版本的SystemElement.java
"""

import math

"""
@author Gianluca Capone & Davide Sgobba
Python转换

此类包含定义计算机公司生产的系统元素的所有变量以及操作这些变量的方法
"""
class SystemElement:
    
    def __init__(self, mod, firm):
        """
        构造函数
        
        Args:
            mod: 系统元素的设计优点 (M-SY_f,t)
            firm: 公司对象引用
        """
        # 变量
        self.mod = mod          # 系统元素的设计优点 (M-SY_f,t)
        self.mu_prog = 0.0      # 系统技术进步分布的均值 (mu-SY_f,t)
        
        # 技术变量和对象
        self.firm = firm        # 访问公司对象
        
    def calc_mod(self):
        """
        根据过去的mod水平和公共知识水平计算系统的mod
        """
        z_max = 0.0
        # 方程14.a
        self.mu_prog = (math.log(self.firm.computer_market.pk_sys) * 
                      (1 - self.firm.computer_market.internal_cum) + 
                      math.log(self.mod) * self.firm.computer_market.internal_cum)
        
        for i in range(1, self.firm.num_of_draws_sys + 1):
            z = math.exp(self.mu_prog + 
                       math.sqrt(self.firm.computer_market.sd_sys) * 
                       self.firm.computer_market.rng.nextGaussian())
            if z > z_max:
                z_max = z
                
        if z_max > self.mod:
            self.mod = z_max
            
    def exit_system(self):
        """
        退出时激活的方法：重置产品层面的最相关变量
        """
        self.mu_prog = 0.0
        self.mod = 0.0 